frontend/device/kindle/device.lua

```diff
@@ -15,10 +15,179 @@ require("ffi/fbink_input_h")
+    end
+end
+
+local function kindleAuthenticateNetwork(essid)
+    local haslipc, lipc = pcall(require, "liblipclua")
+    local lipc_handle
+    if haslipc then
+        lipc_handle = lipc.init("com.github.koreader.networkmgr")
+    end
+    if lipc_handle then
+        lipc_handle:set_string_property("com.lab126.cmd", "ensureConnection", "wifi:" .. essid)
+        lipc_handle:close()
+    end
+end
+
+local function kindleSaveNetwork(data)
+    local haslipc, lipc = pcall(require, "libopenlipclua") -- use our lua lipc library with access to hasharray properties
+    local lipc_handle
+    if haslipc then
+        lipc_handle = lipc.open_no_name()
+    end
+    if lipc_handle then
+        local profile = lipc_handle:new_hasharray()
+        profile:add_hash()
+        profile:put_string(0, "essid", data.ssid)
+        if string.find(data.flags, "WPA") then
+            profile:put_string(0, "secured", "yes")
+            profile:put_string(0, "psk", data.password)
+            profile:put_int(0, "store_nw_user_pref", 0) -- tells amazon we don't want them to have our password
+        else
+            profile:put_string(0, "secured", "no")
+        end
+        lipc_handle:access_hash_property("com.lab126.wifid", "createProfile", profile):destroy() -- destroy the returned empty ha
+        profile:destroy()
+        lipc_handle:close()
+    end
+end
+
+local function kindleGetScanList()
+    local _ = require("gettext")
+    local haslipc, lipc = pcall(require, "libopenlipclua") -- use our lua lipc library with access to hasharray properties
+    local lipc_handle
+    if haslipc then
+        lipc_handle = lipc.open_no_name()
+    end
+    if lipc_handle then
+        if lipc_handle:get_string_property("com.lab126.wifid", "cmState") ~= "CONNECTED" then
+            local ha_input = lipc_handle:new_hasharray()
+            local ha_results = lipc_handle:access_hash_property("com.lab126.wifid", "scanList", ha_input)
+            if ha_results == nil then
+                -- Shouldn't really happen, access_hash_property will throw if LipcAccessHasharrayProperty failed
+                ha_input:destroy()
+                lipc_handle:close()
+                -- NetworkMgr will ask for a re-scan on seeing an empty table, the second attempt *should* work ;).
+                return {}, nil
+            end
+            local scan_result = ha_results:to_table()
+            ha_results:destroy()
+            ha_input:destroy()
+            lipc_handle:close()
+            if not scan_result then
+                -- e.g., to_table hit lha->ha == NULL
+                return {}, nil
+            else
+                return scan_result, nil
+            end
+        end
+        lipc_handle:close()
+        -- return a fake scan list containing only the currently connected profile :)
+        local profile = kindleGetCurrentProfile()
+        return { profile }, nil
+    else
+        return nil, _("Failed to acquire an anonymous lipc handle")
```

Comment:
If we want to show something so technical it probably needs a translator's note.